compat/src/render.js

```diff
@@ -106,10 +106,9 @@ options.vnode = vnode => {
 	if (typeof type != 'function') {
 		// Apply defaultValue to value
 		if (props.defaultValue) {
-			if (!props.value && props.value !== 0) {
+			if (props.value == null && props.value !== 0) {
```

Comment:
Okay sorry about this I was really confused on how to support uncontrolled in this case, so what we want to do is apply `defaultValue` when `props.value` is falsy but not undefined, right?
In the case of us seeing it is controlled (value & onChange present) we apply defaultValue to `props.value` (and delete props.defaultValue)
In the case of uncontrolled we shouldn't enter this branch at all and go for full native dom behavior

Reply:
Sorry to don't explain better about what I did. The issue comes when:

* `defaultValue` -> defined, some string.
* `value` -> undefined
* `onChange` -> defined + is forcing a render (ex: forceUpdate or setState)

In this case, the `props.value` is taking always the value of `defaultValue`, instead of only setting on mounting the component. This causes that although the `value` is undefined, it is acting as controlled... Without the possibility to type anything inside the input text.

This check here doesn't help a lot. Maybe I can change it again to `!props.value`. The approach that I take is on the other file, `src/diff/index.js`. Removing below the `delete props.defaultValue;` then it is possible to check in `src/diff/index.js` that if the user types something different to the `defaultValue`, the `dom.value` can be updated. 

However, I guess that is not the root of the problem and is only a workaround. If you can help me with the familiarity of the correct behavior maybe we can found the root of the problem.

I'm experimenting this without this workaround:

![image](https://user-images.githubusercontent.com/13313058/76014919-bcd3b180-5f1a-11ea-8ebd-1deeeca0576f.png)

After type a "d", the `props.value` is taking the value of `defaultValue`. Then, in the `src/diff/index.js`, the correct `dom.value` is replaced with the `props.value`, when the `props.value` has "bar" (the `defaultValue`).

![image](https://user-images.githubusercontent.com/13313058/76015178-218f0c00-5f1b-11ea-9a97-d4ded3367099.png)

So, the final result is that although I typed a "d", the `dom.value` is resetting again to the defaultValue:

![image](https://user-images.githubusercontent.com/13313058/76015322-5ac77c00-5f1b-11ea-90bf-c96c8aa1d9ea.png)

**After this fix:**

The first part is doing just the same that before (maybe here is the root of the problem that should be fixed):

![image](https://user-images.githubusercontent.com/13313058/76015669-cc072f00-5f1b-11ea-98a2-d01587335847.png)

However, now is not removing `props.defaultValue` with the idea to check then in `src/diff/index.js` if it has defaultValue.

And in `src/diff/index.js` is not updating the `dom.value` because `props.value` is equal to `props.defaultValue` (props.value updated before). And the `dom.value` already has the correct value because is uncontrolled.

And the result is that the input value now is updated correctly:

![image](https://user-images.githubusercontent.com/13313058/76016118-8e56d600-5f1c-11ea-9272-129c8dcfb980.png)

Reply:
Yes that’s what I’m saying, when there is ni value property on props we can assume uncontrolled this means that we can skip this whole condition and rely on the native dom behavior, am I mistaken?

Reply:
Well, I guess that it should enter the condition the first time, when the component is mount. Then I guess that is not necessary.

Reply:
No, never. Uncontrolled should only reflect to the dom-node not to a js-value that’s where we went wrong

Reply:
But entering the first time, then `props.value` is defined, and is entering again...

Reply:
Ahh ok, I am mistaken then! Thank you so much for your time.

Reply:
Hey @aralroca,

so when we talk about controlled and uncontrolled we are mainly talking about JS vs just DOM, we can see that we are talking about DOM when we aren't passing a `value` to an `<input />`. In this case we can rely on the semantics of how `defaultValue` works since under the hood `defaultValue` is reflected correctly on the DOM.
The measure we are referring to here makes it so that this `defaultValue` is also reflected on the JS-prop, for instance this is an `uncontrolled input` in native Preact, where this measure isn't available, as you can see it just works. https://codesandbox.io/s/keen-breeze-39fdh

From this we can conclude that with `compat` we do too much work, we apply `defaultValue` even when it's an uncontrolled input, making it controlled (why your value keeps changing), this should be fixed by checking on line 108 whether we have a defaultValue and if `value` exists as a property on `newProps` this can be done cheaply by saying `!== undefined` or somewhat more expensive by `hasProperty`

Reply:
Thank you so much for your explanation! Your proposal works pretty well. I did a commit in order to fix it!